### ISSUE TITLE
gateway: how to handle separation of envirionemnts? And do we even need to?

### ISSUE THREAD

**Issue by roberts-pumpurs** [equilibriumco/axelar-amplifier-solana#21]
**Summary**

Because Solana program IDs **are** Public keys, it means that to deploy a program (and meaningfully use it), it needs to be deployed to the Solana account associated with the said public key, which can only be achieved by the corresponding private key. Do we need a system for separating the keypairs for devnet and mainnet?

**Motivations**

Separating CI, devnet, mainnet program ids allow us to handle the corresponding Keypair more carefully.

**Final conclusions**

Research issues can contains many information in form of comments and go throw multiple phases till a conclusion/decision is reached. This should be filled only at the end of the research, before closing the issue. Many times, a research issue can derive in other implementation issues, they should be listed here.

**Ready-ready checklist**

- [x] Rationale and requirements are provided (other sections are optional)
- [x] Labels (ticket type and component) are set
- [x] Project & Phase is assigned
- [ ] Estimations are provided
- [ ] Ticket is placed on the timeline
- [ ] 'Needs clarification' label is removed
- [ ] This checklist is removed

**Comment by roberts-pumpurs** [comment]
Extract from Slack:
> Tiago wrote:

The conditionally compiled declare_id!() macro would enable different deployments/environments to have distinct addresses.
I think that mainnet and  devnet should have different secret keys, with different security/access levels.
As for the CI and our local dev environments, using a localnet that allows for maleable account addresses should be sufficient. If not, they could also get a disposable keypair.
In practice, code would look like this:
```
use solana_program::declare_id;
use cfg_if::cfg_if;

cfg_if! {
    if #[cfg(feature = "devnet")] {
        declare_id!("DevnetProgram1111111111111111111111111111111");
    } else if #[cfg(feature = "testnet")] {
        declare_id!("TestnetProgram2222222222222222222222222222222");
    } else if #[cfg(feature = "mainnet")] {
        declare_id!("MainnetProgram3333333333333333333333333333333");
    } else {
        // Default ID if no feature is specified
        declare_id!("DefaultProgram4444444444444444444444444444444");
    }
}
```


> Roberts wrote:

I think this requires more research and consideration & should be discussed in a gh issue. Because creating something like this would enforce every 3d party program, which wants to integrate with Solana gateway, to also propagate these feature flags for devnet & mainnet. While it makes sense in terms of separating the environments, Id prefer to read & research how this is handled by other solana program developers

**Comment by eloylp** [comment]
> I think this requires more research and consideration & should be discussed in a gh issue. Because creating something like this would enforce every 3d party program, which wants to integrate with Solana gateway, to also propagate these feature flags for devnet & mainnet. While it makes sense in terms of separating the environments, Id prefer to read & research how this is handled by other solana program developers

I agree with this. And i think we need to consider this before going for mainnet. Moving this issue to P6

**Comment by frenzox** [comment]
So, the private key associated with the program id is only required for the [initial deployment](https://solana.com/docs/programs/deploying#initial-deployment). When the initial deployment is performed, the upgrade authority is set. It can be passed as a command line argument (`--upgrade-authority`) or, if not set, it uses the default configured keypair for the solana cli. After that, the program is upgraded using the upgrade authority as signer.

I would strongly suggest using the same `program_id` for all clusters: devnet, testnet and mainnet. If we look at the metadata, token programs, etc. none have a different ids for different clusters. Honestly, I think that would be a bit madness. As @roberts-pumpurs rightfully pointed out, this would be propagated to crates depending on our programs, and they would need to implement some similar configuration and that would be propagated until the entrypoint crate.

We do need some proper way to handle keys. We need 1 keypair for each program (that's going to be used across the different clusters, the pubkey being the `program_id`) and 3 keypairs, one for each cluster, to be used as upgrade authority (and that's also the `Owner` for some contracts like ITS).

**Comment by frenzox** [comment]
@eloylp Having said that, I don't think there's much more to do in the context of this issue. I guess we have some other issue and likely requirements on how to deal with keys and this should probably be taken care in that context now. Let me know what you think.

**Comment by eloylp** [comment]
> [@eloylp](https://github.com/eloylp) Having said that, I don't think there's much more to do in the context of this issue. I guess we have some other issue and likely requirements on how to deal with keys and this should probably be taken care in that context now. Let me know what you think.

I agree @frenzox  - it makes sense- We only care about upgrade auhtorities and that frees us to preserve same program ids on every enviornment. Thanks for your great comment. Closing
